### PR TITLE
Do not display layer in legend when outside of map frame

### LINKED ISSUE #38402: Layout Legend : filter out Rasters by visibility and extent
**Feature description.**
Currently only features are filtered out by the available option.

It would be nice to not display the values for a raster that is not visible and/or not within the atlas.

### PR DESCRIPTION
**Pull request by roya0045**
## Description

Fixes #38402

This enable the maphittest for raster layer, this will prevent displayed unpresent layer when legend filtering is enabled.

The same behaviour could easily be extended to meshes and point cloud layers.

Been sitting on this one until I could test. Works as intended.

@nyalldawson should I add a test?

### CODE REVIEW

**Comment by roya0045** [comment]
waiting for review.

**Comment by roya0045** [comment]
Awaiting review.

File: src/core/qgsmaphittest.cpp [hunk 1]
```diff
@@ -205,3 +206,17 @@ void QgsMapHitTest::runHitTestLayer( QgsVectorLayer *vl, SymbolSet &usedSymbols,
...
 }
 
+bool QgsMapHitTest::rasterVisible( QgsRasterLayer *layer ) const
+{
+  if ( ! layer->dataProvider() )
+    return false;
+  QgsRectangle footprint = layer->dataProvider()->extent();
+  if ( mSettings.destinationCrs() != layer->crs() )
+  {
+    QgsCoordinateTransform ct = QgsCoordinateTransform( mSettings.destinationCrs(), layer->crs(), mSettings.transformContext() );
```
Comment (uclaros) [inline src/core/qgsmaphittest.cpp:216]: Shouldn't source be the `layer->crs()` and target the `mSettings.destinationCrs()`?
Your test could handle a case of crs transformation to be sure.

File: src/core/qgsmaphittest.cpp [hunk 2]
```diff
@@ -205,3 +206,17 @@ void QgsMapHitTest::runHitTestLayer( QgsVectorLayer *vl, SymbolSet &usedSymbols,
...
+    footprint = ct.transformBoundingBox( footprint );
```
Comment (uclaros) [inline src/core/qgsmaphittest.cpp:217]: Is it not possible for this to throw a `QgsCsException` that should be handled?

File: src/core/qgsmaphittest.h [hunk 3]
```diff
@@ -71,6 +72,13 @@ class CORE_EXPORT QgsMapHitTest
      */
     bool legendKeyVisible( const QString &ruleKey, QgsVectorLayer *layer ) const;
 
+    /**
+     * Tests whether a given legend key is visible for a specified layer.
```
Comment (uclaros) [inline src/core/qgsmaphittest.h:76]: This implies that raster classes are also handled just like for the vector layers. 
```suggestion
     * Tests whether a raster layer is visible.
```

**Review by uclaros — COMMENTED** [review]
Raster classes, nodata and total transparency could be handled in the future to feature-match vector layers, but layer scale visibility should probably be handled by this pr.

File: src/core/qgsmaphittest.cpp [hunk 4]
```diff
@@ -205,3 +206,17 @@ void QgsMapHitTest::runHitTestLayer( QgsVectorLayer *vl, SymbolSet &usedSymbols,
...
+  }
+  if ( mSettings.extent().intersects( footprint ) )
```
Comment (uclaros) [inline src/core/qgsmaphittest.cpp:219]: I'm not too familiar with the class, but shouldn't `visibleExtent()` be used instead?

Comment (nyalldawson) [reply src/core/qgsmaphittest.cpp:219]: Yes, definitely

Comment (roya0045) [reply src/core/qgsmaphittest.cpp:219]: There is also 'vibiblePolygon()', doc states it can be rotated, unsure if this is better.

**Comment by roya0045** [comment]
@uclaros Thanks for the review! I'll try to address this soon.

File: src/core/qgsmaphittest.cpp (hunk 2 above)
Comment (roya0045) [reply src/core/qgsmaphittest.cpp:217]: Yes. though it's not currently done for the vector hit test.

**Comment by roya0045** [comment]
comments have been addressed, awaiting further review or merge.

**Comment by roya0045** [comment]
Not stale.

**Comment by roya0045** [comment]
fine

**Comment by roya0045** [comment]
still ok.

**Comment by roya0045** [comment]
:(

**Comment by roya0045** [comment]
awaiting review

**Comment by roya0045** [comment]
same

**Comment by roya0045** [comment]
staying alive

**Comment by roya0045** [comment]
ok

**Comment by roya0045** [comment]
ok

**Comment by roya0045** [comment]
Not stale.

**Comment by roya0045** [comment]
Still good, @lbartoletti  Thanks Loic for destaling my other PRs!

**Comment by roya0045** [comment]
@nyalldawson I made it more general, should be more useful.

**Comment by roya0045** [comment]
still fine

**Comment by roya0045** [comment]
@nyalldawson Should I recreate a new PR or reopening this one would be fine?

**Pull request closed without merging**